File: worker/marvin/room/protocol.py
"""Wire protocol between the worker and the web clients. Mirrors web/src/protocol.ts."""
from __future__ import annotations

import json
from typing import Any

TOPIC_EVENTS = "marvin"  # worker -> everyone
TOPIC_CONTROL = "marvin-control"  # clients -> worker
AGENT_IDENTITY = "marvin"
MAX_PACKET = 12_000  # LiveKit reliable packets cap at 15 KiB; leave headroom


def encode(event: dict[str, Any]) -> bytes:
    return json.dumps(event, ensure_ascii=False).encode()


def decode(payload: bytes) -> dict[str, Any]:
    return json.loads(payload.decode())


def split_text_event(event: dict[str, Any]) -> list[dict[str, Any]]:
    """A long `text` event becomes many `text_delta` events so each packet fits."""
    if event.get("kind") != "text" or len(encode(event)) <= MAX_PACKET:
        return [event]
    text = event["text"]
    step = MAX_PACKET // 2
    return [{"kind": "text_delta", "text": text[i : i + step]} for i in range(0, len(text), step)]

File: web/src/protocol.ts
// Wire protocol between the Marvin worker and the room. Mirrors worker/marvin/room/protocol.py.
export const TOPIC_EVENTS = "marvin"; // worker -> everyone
export const TOPIC_CONTROL = "marvin-control"; // clients -> worker
export const AGENT_IDENTITY = "marvin";

export type MarvinEvent =
  | { kind: "app_links"; links: { label: string; url: string }[] } // where the room's app is served (from rooms.yaml)
  | { kind: "status"; state: "idle" | "thinking" | "waiting_approval" }
  | { kind: "transcript"; speaker: string; text: string; start: number; end: number; at?: number; final?: boolean }
  | { kind: "turn_start"; asked_by: string; question: string }
  | { kind: "text_delta"; text: string }
  | { kind: "text"; text: string }
  | { kind: "tool_use"; id: string; tool: string; input: Record<string, unknown> }
  | { kind: "tool_result"; id: string; output: string; is_error: boolean }
  | { kind: "permission_request"; id: string; tool: string; input: Record<string, unknown> }
  | { kind: "permission_resolved"; id: string; allow: boolean; by: string }
  | { kind: "auto_approve"; on: boolean; by: string } // "always allow" was toggled by someone in the room
  | { kind: "result"; subtype: string; is_error: boolean; cost_usd: number | null; duration_ms: number | null; session_id: string | null }
  | { kind: "attachment"; name: string; path: string; by: string } // an image landed on disk, waiting for the next turn
  | { kind: "denied"; action: string; by: string; reason: string } // a control message was refused (e.g. auto_approve without the admin role)
  | { kind: "error"; message: string };

export type ControlMessage =
  | { action: "approve" | "deny"; id: string }
  | { action: "auto_approve"; on: boolean } // stop asking for every tool until turned back off
  | { action: "ask"; text: string } // typed fallback: same path as a spoken "Marvin, ..."
  | { action: "image"; id: string; seq: number; total: number; mime: string; name: string; data: string } // base64, split across packets
  | { action: "interrupt" };

export const IMAGE_CHUNK = 8_000; // base64 chars per packet, well under LiveKit's 15 KiB cap
export const MAX_IMAGE_BYTES = 8_000_000;
